gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestParamsBuilder.java

```diff
@@ -56,20 +59,44 @@ public static RequestParamsBuilder create() {
    * pre-configured path templates. This method is called repeatedly for each configured routing
    * rule parameter, it's possible that the incoming field value from request is null or there is no
    * matches found, we'll continue the match-and-extract process for the next routing rule parameter
-   * in such case.
+   * in such case. This method will percent encode both the header key and header value.
    *
    * @param fieldValue the field value from a request
    * @param headerKey the header key for the routing header param
    * @param pathTemplate {@link PathTemplate} the path template used for match-and-extract
    */
   public void add(String fieldValue, String headerKey, PathTemplate pathTemplate) {
-    if (fieldValue == null) {
+    if (fieldValue == null || fieldValue.isEmpty()) {
       return;
     }
     Map<String, String> matchedValues = pathTemplate.match(fieldValue);
     if (matchedValues != null && matchedValues.containsKey(headerKey)) {
-      paramsBuilder.put(headerKey, matchedValues.get(headerKey));
+      String encodedKey = percentEncodeString(headerKey);
+      String encodedValue = percentEncodeString(matchedValues.get(headerKey));
+      paramsBuilder.put(encodedKey, encodedValue);
+    }
+  }
+
+  /**
+   * Add an entry to paramsBuilder with key-value pairing of (headerKey, fieldValue). The only
+   * validation done is to ensure the headerKey and fieldValue are not null and non-empty. This
+   * method will percent encode both the header key and header value.
+   *
+   * @param headerKey the header key for the routing header param
+   * @param fieldValue the field value from a request
+   */
+  public void add(String headerKey, String fieldValue) {
```

Comment:
Now that we know we have to encode the headers, I think we should do it in a more generic way, like in `RequestUrlParamsEncoder` or even encode all `extraHeaders`.

Reply:
>Do we need to encode all headers?

That's a great question. I think we should but I'm not sure. There are some security risks of not encoding urls, so I was under the impression that headers should be encoded and are already encoded as well by the downstream grpc or http libraries. Probably not part of this PR but we can do a little more investigation on the headers encoding in general. As part of this PR, I think we should move this logic to `RequestUrlParamsEncoder` because that class is specific for the routing headers and the name is perfect for this logic.

Reply:
>Do we need to encode all headers?

That's a great question. I think we should but I'm not sure. There are some security risks of not encoding urls, so I was under the impression that headers should be encoded and are already encoded as well by the downstream grpc or http libraries. Probably not part of this PR but we can do a little more investigation on the headers encoding in general. As part of this PR, I think we should move this logic to `RequestUrlParamsEncoder` because that class is specific for the routing headers and the name is perfect for this logic.

Reply:
> Probably not part of this PR but we can do a little more investigation on the headers encoding in general.

Sure! I will create an issue in the backlog to investigate this. 

> As part of this PR, I think we should move this logic to RequestUrlParamsEncoder because that class is specific for the routing headers and the name is perfect for this logic.

Yep, I agree. The only concern I have is that changing `RequestUrlParamsEncoder` to now encode the header values might be a behavior breaking change: https://github.com/googleapis/sdk-platform-java/blob/ef4a7ff91b259f0b650cc06bd664eadca7bd17c9/gax-java/gax/src/main/java/com/google/api/gax/rpc/RequestUrlParamsEncoder.java#L68-L79 as it expects the params to the percent encoded already.

Reply:
This class is only used for encoding the request params(which is effective routing headers), and it is marked as `@InternalAPI` so we should be able to change it at anytime.

Reply:
Sounds good. I'll migrate the encoding logic here.

Reply:
@blakeli0 I took out the logic for doing validation inside this class. It's an `@InternalApi` and only used by the {Transport}{Request_Type}RequestParamCallable.

Let me know you're fine with that. Otherwise I can revert that part of the change.

Reply:
SGTM.

Reply:
@blakeli0 @lqiu96 Did this effectively change the behavior of implicit routing headers from not encoding params to now URL encoding them?

Reply:
Yes, it did. There was another [discussion](https://github.com/googleapis/sdk-platform-java/pull/1667#discussion_r1199286063) regarding should we add it or not for existing routing headers, we discussed offline and confirmed with Slava that all routing header should be encoded.

Reply:
Yep, there was a discussion with Slava about the need to encode both the header key and value. We saw there was customer issue (b/283866374) regarding having an unencoded value.